### PR TITLE
[Auto Parallel] Improve the c++ dist attr

### PR DESCRIPTION
**Pull request by aoyulong**
### PR types

Others

### PR changes

Others

### Describe

This pr improves the c++ distributed attributes.
- Fix some bugs
- Support serialization of dist attrs
- Polish the attr of var_desc

### CODE REVIEW

**Review by chenwhql — DISMISSED** [review]
LGTM

**Review by XiaoguangHu01 — APPROVED** [review]
LGTM

File: python/paddle/fluid/framework.py [hunk 1]
```diff
@@ -2612,7 +2612,7 @@ def attr_names(self):
         """Get the names of all attributes defined."""
         return self.desc.attr_names()
 
-    def _get_attr(self, name):
+    def attr(self, name):
```
Comment (jzhang533) [inline python/paddle/fluid/framework.py:2615]: after consulting @phlrain 
this change is OK for fluid API deprecation plan

**Review by jzhang533 — APPROVED** [review]
LGTM

**Pull request merged**